sys/include/xtimer.h

```diff
@@ -0,0 +1,442 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup  sys_xtimer Timers
+ * @ingroup   sys
+ * @brief     Provides a high level timer module to register
+ *            timers, get current system time, and let a thread sleep for
+ *            a certain amount of time.
+ *
+ * The implementation takes one low-level timer that is supposed to run at 1MHz
+ * speed and multiplexes it.
+ *
+ * Insertion and removal of timers has O(n) complexity with (n) being the
+ * number of active timers.  The reason for this is that multiplexing is
```

Comment:
> You're saying you lied in your thesis?

No, but xtimer never guaranteed realtime capabilities.

> Seriously, I know that there are several places in RIOT that breaks real-time guarantees, but at least for the core components we should a) try not to introduce new RT-breaking parts (particular not silently) and b) repair the broken stuff.

I'd say, real-time timers are just not the scope of xtimer. All the multiplexing while trying to save memory is too complex. Artificially limiting xtimer to "n" timers and call it realtime capable because of that will make real-time interested users laugh and look away, while causing trouble for users running in weird problems because the timers run out.

There are a bunch of void returning functions that never fail. While having a default of no limiting could be implemented, it would mean having to return success/failure when setting timers, and worse, having to check for those return values wherever timers are used.

Would you be OK if the ducomentation says, "if you need realtime timers, use periph/timers directly"?